### PR TITLE
[GEN-1196] Parameterize docker images

### PR DESCRIPTION
**Pull request by rxu17**
**Purpose:** This is a draft PR waiting on a bug due to the GN annotator. This PR allows us to input specific docker images for each of our process steps. Previously, we had to manually edit the `container` variable in each of the modules to use a different image to test it locally. This wasn't possible in Nextflow Tower where we had to create a separate branch in nf-genie to use different containers which makes our workflow difficult. Now one should be able to specify the docker image for a specific process if they want to use one that's not the default.

I looked through various nf-core repos and I wasn't able to find one that used parameterized docker image inputs and had the use case that we needed. I thought about using profiles in the nextflow.config but we would still need to specify the specific image then and I don't think it offers the flexibility we would like.

Here we are able to utilize the different purposes set by nextflow.config and nextflow_schema.json.

- changes to `nextflow_schema.json` are for the nextflow tower interface and visualizing the parameters and their defaults there
- changes to `nextflow.config` are for setting defaults for production vs test profiles when running the pipeline locally

**Testing:**
- Tested on validation, processing (mutation and non-mutation), public release and create_data guide

### CODE REVIEW

File: nextflow_schema.json [hunk 1]
```diff
@@ -37,6 +37,26 @@
                     "description": "Release name (specify TEST.consortium/TEST.public to test consortium and public releases). E.g: 13.1-consortium",
                     "default": "TEST.consortium",
                     "pattern": "\\d+[.]\\d+-(consortium|public)$|TEST.public|TEST.consortium"
+                },
+                "main_pipeline_docker":{
+                    "type": "string",
+                    "description": "Docker used in main processes - validate_data, process_main, process_maf, create_consortium_release, and create_public_release",
+                    "default": "sagebionetworks/genie:latest"
```
Comment (thomasyu888) [inline nextflow_schema.json:44]: Should we use the ghcr image?

**Comment by BWMac** [comment]
@rxu17 and I chatted over Slack and I made a couple of suggestions. She is going to further explore potentially using `profiles` and the way the default `params` are set. I'm happy to circle back to this for more review once that exploration is complete.

**Comment by danlu1** [comment]
@rxu17 this might be of interest: https://www.nextflow.io/docs/latest/container.html#:~:text=to%20true.-,Multiple%20containers,Read%20the%20Process%20scope%20section%20to%20learn%20more%20about%20processes%20configuration.,-Apptainer%20%26%20Docker%20Hub

**Comment by rxu17** [comment]
After the suggestion by @BWMac , I've moved the parameter definitions into the two profiles we have so we have less docker containers we have to manually define when we run on develop vs prod. For all other situations (feature branches) that require containers defined outside of our two profiles, we can define the specific parameter in the command line call (when running locally)

File: nextflow_schema.json (hunk 1 above)
Comment (rxu17) [reply nextflow_schema.json:44]: The GHCR image hasn't been built off `main` yet until we do a release. I think we can create another PR to update this after that's done.

**Comment by rxu17** [comment]
> @rxu17 this might be of interest: https://www.nextflow.io/docs/latest/container.html#:~:text=to%20true.-,Multiple%20containers,Read%20the%20Process%20scope%20section%20to%20learn%20more%20about%20processes%20configuration.,-Apptainer%20%26%20Docker%20Hub

Thanks for the suggestion! I looked into this prior and unfortunately this specific method doesn't seem to offer us much flexibility in terms of overriding the parameter values both locally and on Tower because it's not actually creating parameters but defining the containers directly.

**Comment by thomasyu888** [comment]
> > @rxu17 this might be of interest: https://www.nextflow.io/docs/latest/container.html#:~:text=to%20true.-,Multiple%20containers,Read%20the%20Process%20scope%20section%20to%20learn%20more%20about%20processes%20configuration.,-Apptainer%20%26%20Docker%20Hub
> 
> Thanks for the suggestion! I looked into this prior and unfortunately this specific method doesn't seem to offer us much flexibility in terms of overriding the parameter values both locally and on Tower because it's not actually creating parameters but defining the containers directly.

@rxu17 have you tried setting this with the different profiles?

**Comment by rxu17** [comment]
> @rxu17 have you tried setting this with the different profiles?
Setting the parameters we have for docker? Or the `container` directive?

**Comment by thomasyu888** [comment]
> > @rxu17 have you tried setting this with the different profiles?
> 
> Setting the parameters we have for docker? Or the `container` directive?

The container directive. Can you set the `withName: process` within different profiles?

**Comment by rxu17** [comment]
> The container directive. Can you set the withName: process within different profiles?
Yes. 

I think you are asking if we can do this:
```
// nextflow.config
profiles {
    aws_test {
        params {
            main_pipeline_docker = 'image_name'
            create_data_guide_docker = 'image_name'
        }
        process {
            withName:validate_only {
                container = params.main_pipeline_docker
            }
            withName:main_process {
                container = params.main_pipeline_docker
            }
        }
    }
}

// validate_only.nf
process validate_only {
    script:
    """
    # Process validate_only script
    """
}
```

VS

this is what we currently have in place with this PR:
```
// nextflow.config
profiles {
    aws_test {
        params {
            main_pipeline_docker = 'image_name'
            create_data_guide_docker = 'image_name'
        }
    }
}

// validate_only.nf
process validate_only {
    container: params$main_pipeline_docker
    script:
    """
    # Process validate_only script
    """
```
Is that right?

**Comment by thomasyu888** [comment]
> Is that right? I think you can do the top option (would have to try). It does slightly improve maintainability in that if the docker param names change in the future/adding new docker image params, we don't have to modify the modules just the nextflow.config.

Ah, I see the benefit to having the docker containers be parameters, originally I had thought:

```
// nextflow.config
profiles {
    aws_test {
        process {
            withName:validate_only {
                container = "ghcr.io/..."
            }
            withName:main_process {
                container = "ghcr.io/..."
            }
        }
    }
}

// validate_only.nf
process validate_only {

    script:
    """
    # Process validate_only script
    """
}
```

But then I realized you would need to keep adding profiles every time you built and wanted to test an image.  I'll defer to you and the rest of the team on what they think, but I think you came up with a great solution!  Thanks!

File: nextflow_schema.json [hunk 2]
```diff
@@ -13,7 +13,6 @@
             "properties": {
                 "process_type": {
                     "type": "string",
-                    "default": "only_validate",
```
Comment (thomasyu888) [inline nextflow_schema.json:16]: Will this be an issue for the orca dag that triggers daily validation?

Comment (rxu17) [reply nextflow_schema.json:16]: Great catch, I'm not sure how that got deleted. 

The good thing is changes in the `nextflow_schema.json` will not affect the code execution in actuality (this param is initialized to `only_validate` in `main.nf`) but only the aesthetic appearance of what you see on Nextflow Tower. So deleting this default just makes the param box appear as a blank when you click `Launch` for a pipeline.

**Review by BWMac — APPROVED** [review]
Looks great! 

Not sure if it would fit with the flexible strategy you are going for here, but in the future if you want you could look into using [labels](https://www.nextflow.io/docs/latest/process.html#label). They allow you you set containers for processes along with the compute resource allocations at the same time. Or you could even mix-and-match resource and container labels.

File: nextflow.config [hunk 3]
```diff
@@ -16,6 +16,13 @@ profiles {
 			memory = 16.GB
 			cpus = 4
 		}
+		params {
+			// docker image parameters, see nextflow_schema.json for details
+			main_pipeline_docker = "sagebionetworks/genie:develop"
+			main_release_utils_docker = "sagebionetworks/main-genie-release-utils"
+			find_maf_artifacts_docker = "sagebionetworks/genie-artifact-finder"
+			create_data_guide_docker = "sagebionetworks/genie-data-guide"
```
[the remark this answers is not in the document]
Comment (BWMac) [reply nextflow.config:24]: Ah, that makes sense. Sounds good :)

**Review by BryanFauble — APPROVED** [review]
Cool to see items like this being extracted out to a configuration layer!

Concepts like this might be what will enable us to use a plugin architecture to re-use some of these components.

**Review by thomasyu888 — APPROVED** [review]
🔥 LGTM!

**Pull request merged**